### ISSUE TITLE
Patches not building on RT kernel

### ISSUE THREAD

**Issue by justablob** [sfjro/aufs-standalone#10]
Good evening

When trying to build aufs5 on a preempt-rt kernel, the build fails. I'm using Linux with preempt-rt version `5.15.26-rt34` and aufs `aufs5.15.5`. Here is the error:
```
fs/aufs/i_op.c: In function ‘au_pin_hdir_set_owner’:
fs/aufs/i_op.c:639:52: error: ‘struct rw_semaphore’ has no member named ‘owner’
  639 |         atomic_long_set(&p->hdir->hi_inode->i_rwsem.owner, (long)task);
      |                                                    ^
make[3]: *** [scripts/Makefile.build:277: fs/aufs/i_op.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [scripts/Makefile.build:540: fs/aufs] Error 2
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:1868: fs] Error 2
make[1]: Leaving directory '/root/buildroot-2022.02-rc3/output/build/linux-custom'
```

The build succeeds with only preempt-rt or only aufs, but with both of them, it fails like that.
The host is a Debian Bullseye amd64 virtual machine. I'm using buildroot to automate the build, since doing it all manually is a pain.

**Comment by paulgortmaker** [comment]
You need to apply the rt.patch as well as base/kbuild/standalone patches.

https://github.com/sfjro/aufs5-standalone/blob/aufs5.15.5/rt.patch

Note that the rt.patch does not really need to be kept as a standalone item - it could/should be folded into the fs/aufs/i_op.c unconditionally for anything v5.3 and newer.   This is because the Kconfig for PREEMPT_RT was added in mainline linux a50a3f4b6a31

**Comment by paulgortmaker** [comment]
Actually as it turns out -- the rt.patch does need to be updated for v5.15 kernels..  It needs to bounce through rwbase now.  Please see:

https://lists.yoctoproject.org/g/linux-yocto/message/11083

for details.

**Comment by sfjro** [comment]
Paul Gortmaker:
> Actually as it turns out -- the rt.patch does need to be updated for v5.15 kernels..  It needs to bounce through rwbase now.  Please see:
>
> https://lists.yoctoproject.org/g/linux-yocto/message/11083
>
> for details.

Thanx for the patch.
I will put it into aufs releases as 'rt-5.15.7.patch'.

By the way, on 09 Mar 2022, I have replied your message. Was it
delivered? I cannot see it on
https://github.com/sfjro/aufs5-standalone/issues/10. I don't know why.


J. R. Okajima

**Comment by justablob** [comment]
@sfjro I don't think your message was delivered. Either way, I'm gonna wait on the 5.17 release. I'm not in a hurry so take your time :)

**Comment by paulgortmaker** [comment]
I did not see any message from March 9th - just the one from ~24h ago.

With regards to other AUFS patches - You might also find these links of interest:

https://lists.yoctoproject.org/g/linux-yocto/message/11084

https://lists.yoctoproject.org/g/linux-yocto/message/11085

**Comment by sfjro** [comment]
Paul Gortmaker:
> I did not see any message from March 9th - just the one from ~24h ago.

Ok, my message was totally gone somewhere. But it is ok since the
message becomes less important now.


> With regards to other AUFS patches - You might also find these links of interest:
>
> https://lists.yoctoproject.org/g/linux-yocto/message/11084
> https://lists.yoctoproject.org/g/linux-yocto/message/11085

Hmm, this patch tries supporting the case of "patched but disabled".
That is interesting since I would say "if you disable it, why did you
apply the patch?" for such case. But now I think I could understand such
case. I am going to apply your patch to aufs release.

Thank you.

J. R. Okajima

**Comment by sfjro** [comment]
------- Blind-Carbon-Copy

From: "J. R. Okajima" ***@***.***>
To: ***@***.***
Cc: ***@***.***, ***@***.***
Subject: aufs5 GIT release
MIME-Version: 1.0
Content-Type: text/plain; charset="us-ascii"
Content-ID: ***@***.***>
Date: Mon, 04 Apr 2022 00:43:36 +0900
Message-ID: ***@***.***>

o news
- - a patch for RT v5.15.7 and later is added by Paul Gortmaker.
- - aufs5-mmap.patch is updated to support the case where aufs is disabled
  by Paul Gortmaker.
  Note that I didn't test the patch with aufs is disabled, just
  compiled.


J. R. Okajima

- ----------------------------------------

- - aufs5-linux.git
      aufs mmap: reduce overhead for"code present but disabled" use case.

- - aufs5-standalone.git
  ditto

- - aufs5-standalone.git#aufs5.15.5..aufs5.x-rcN
  Addition to above,
      aufs standalone: rt-5.15.7.patch for RT kernel v5.15.7-rt23 and later

- - aufs-util.git
  nothing

------- End of Blind-Carbon-Copy

**Comment by sfjro** [comment]
Hello The Blobbo,

> Here is the error:=0D
> ```=0D
> fs/aufs/i_op.c: In function =E2=80=98au_pin_hdir_set_owner=E2=80=99:=0D
> fs/aufs/i_op.c:639:52: error: =E2=80=98struct rw_semaphore=E2=80=99 has n=
> o member named =E2=80=98owner=E2=80=99=0D
>   639 |         atomic_long_set(&p->hdir->hi_inode->i_rwsem.owner, (long)=
> task);=0D
>       |                                                    ^=0D

There is a patch for this issue.
Try aufs5-standalone.git/rt.patch.


J. R. Okajima

**Comment by floion** [comment]
Hi @sfjro would it be possible to include the RT patch in newer aufs releases?

**Comment by sfjro** [comment]
Hello Florin,

Florin Sarbu:
> Hi @sfjro would it be possible to include the RT patch in newer aufs releases?

Which kernel version?
When I have time, I'll try.
But this summer heat is driving me crazy, I feel like smashing my PC
with a sledgehammer.
Don't expect me too much.


J. R. Okajima
